feat(iota-framework)!: Add create_auth_info_v1 function

TheMrAI opened the pull request:
# Description of change

Introduce `create_auth_info_v1` and `create_auth_info_self_v1` native functions.
These functions provide the capability of creating an `AuthenticatorInfoV1` pointing to a desired `authenticate`
function.

## How the change has been tested

With the temporary `account_auth_example` unit tests:

```
cargo run -p iota -- move test -p account_auth_example
```

Unfortunately the features aren't fully tested as it isn't entirely clear how they could be. The newly introduced
verifier is more or less fully tested, with `create_auth_self_v1`.

But even this is a limited test as it doesn't mimic proper deployment scenarios.
Both `create_auth_info_v1` and `create_auth_info_self_v1` may be only tested in e2e tests as the currently available testing frameworks aren't necessarily prepared for these scenarios.

## Links to any relevant issues

fixes #8117 
fixes #8415 

### Release Notes

- [x] Protocol: Introduce new move native function `create_auth_info_v1` and `create_auth_info_self_v1` with its gas costs.
- [ ] Nodes (Validators and Full nodes):
- [ ] Indexer:
- [ ] JSON-RPC:
- [ ] GraphQL:
- [ ] CLI:
- [ ] Rust SDK:
- [ ] REST API:

TheMrAI committed:
Add basic framework for create_auth_info_v1 function

account_auth_example and modifications within external-crates/move/crates/move-vm-runtime/src/loader.rs
are only there so that we could quickly test if create_auth_info_v1 would work in
this setup, but will get removed later.

miker83z commented on crates/iota-framework/packages/iota-framework/sources/account.move:
```diff
@@ -0,0 +1,24 @@
+public struct AuthenticatorInfoV1 has store, drop {
+  // this is actually storage id
+  package: address,
+  module_name: String, // `module` is a keyword in move and the parser unfortunately isn't smart enough to realize that one couldn't start a module declaration within a structure, or function argument etc.
```
Alright, could you then also make it coherent for the function? I.e., `function_name`?

TheMrAI replied:
Of course of course, do not a worry. We experimenting here.

valeriyr replied:
I don't think it looks better, but Move has a feature that allows using a keyword as an identifier. Something like this:
```
`module`: String,
```

miker83z commented on crates/iota-framework/packages/iota-framework/sources/account.move:
```diff
@@ -0,0 +1,24 @@
+#[allow(unused_field)]
+public struct AuthenticatorInfoV1 has store, drop {
+  // this is actually storage id
+  package: address,
```
What are the implications of making this an `ID` instead of an `address`?

TheMrAI replied:
Haven't considered it. Followed the specifications so far.

miker83z replied:
Could you please check? We experimenting here. 😉

TheMrAI committed:
Implement verify_authenticate_func

While the verifier is functionally complete, the testing for it isn't.
The current tests rely on the hacks from the previous changes where
we allowed the lookup of functions from the current package.
This case in reality is impossible to write down, but it allowed
us to write the tests.
Furthermore any real test depends on the implementation of proper
module load functionality being available from NativeContext,
which may not be trivial to write.

miker83z commented on iota-execution/latest/iota-verifier/src/account_auth_verifier.rs:
```diff
@@ -0,0 +1,141 @@
+        )));
+    }
+
+    if function_definition.visibility != Visibility::Public {
```
Not for this first iteration of development, but we should consider whether it's worth relax this constraint to allow public and private:
- we can execute private functions from the rust side
- a dev, setting this function as private, means that this is declared as not taking part to composability with other `authenticate()` (same logic as using just `entry` for normal functions)

TheMrAI replied:
I was thinking about this as well, but couldn't really figure out what would make sense, so landed on leaving it public.
Didn't want to potentially break some move invariants. For example if I allow private functions to be called, I need to somehow also enforce that
the `create_auth_info` call originates from the same module as the targeted `authenticate` is in.

On your first point are you referring to dev mode Tx execution or what exactly? Then it could make sense.

TheMrAI replied:
Created a sub issue to track this: https://github.com/iotaledger/iota/issues/8215

miker83z commented on iota-execution/latest/iota-verifier/src/account_auth_verifier.rs:
```diff
@@ -0,0 +1,141 @@
+        .iter()
+        .take(function_signature.len() - 2)
+    {
+        if token.is_mutable_reference() {
```
It is a bit more complex. It must be either a primitive type or just an immutable reference.
- `is_mutable_reference()` blocks `MutableReference` but not `Datatype`, `DatatypeInstantiation` and `TypeParameter` of objects -> by value.
- something like checking if the token is of a primitive value first and then, if not, it can only be a `Reference` is needed https://github.com/iotaledger/iota/blob/4660e9505235ecd54e9a5e02fdaade071622cfaf/crates/iota-json/src/lib.rs#L698

TheMrAI replied:
Ahh, very good catch thank you! I remember seeing those, but forgot to investigate further.

TheMrAI replied:
Fixed in a659ac6109d027024f1995e0e0d2eb08bec0b805 
with the caveat that I didn't restrict it as much at all.
Made a ticket for figuring this out: https://github.com/iotaledger/iota/issues/8216
I suspect we can only really inspect the consequences when we have a running setup.

miker83z commented on iota-execution/latest/iota-verifier/src/account_auth_verifier.rs:
```diff
@@ -0,0 +1,141 @@
+    // TxContext can only be an immutable reference. Passing it as mutable would
+    // allow `authenticate` functions to create objects, which would be
+    // problematic.
+    if let SignatureToken::Reference(referent) = tx_context {
```
We currently have a `TxContext::kind()` function similar to this, but allowing also mutable reference. It would be nice to add a new function there just for immutable ref and then copy both for the `AuthContext`.

TheMrAI replied:
Uhh very good thank you!

TheMrAI replied:
Fixed in a659ac6109d027024f1995e0e0d2eb08bec0b805

TheMrAI committed:
Verifier now properly tests AuthContext/TxContext types

Furthermore verifier check for other arguments have been covered with
additional tests. The types remain as flexible as they can be for now,
until we are sure that they need more restrictions.
Even so the tests will be there to check these.

TheMrAI committed:
Update SignatureToken docs with some examples and better description

It may confuse the uniniciated that SignatureToken's can contain
unresolved template types.
These examples can clarify that the types aren't resolved on this
level yet.

TheMrAI committed:
Change AuthenticatorInfoV1 packege field to ID

Previously it was an address, but on the Rust side we would use it
as an ID. Furthermore a package in IOTA os an object, so referring
to it as an ID should be technically correct.
At this point it isn't visible if this may cause issues, but for
now we are going with it.

miker83z commented on iota-execution/latest/iota-verifier/src/account_auth_verifier.rs:
```diff
@@ -0,0 +1,133 @@
+    // functions, so similarly to TxContext, it is simply not allowed.
+    if !matches!(
+        AuthContext::kind(module, auth_context),
+        AuthContextKind::Immutable
```
To provide more detailed errors, this could match all cases and throw 2 different errors for `None` and `Mutable`. Same for TxContext below.

TheMrAI replied:
If this errors, that means that the code failed during execution.
The user had to write some code, deploy it, then call it from some transaction (which is probably another deployed contract).
If it fails at this point, I am not sure how they are helped with more detail.

Ideally what should happen is that during verification the code collects all errors and reports them as one,
so the user may fix all of them at once.
But this is more of a compiler level function not a runtime/pre-runtime bytecode safety check as others also seem to be.

Here at the verifier stage the user is beyond easy help. We don't know which line of code failed, what spans etc.

Given that for the `view` function the council wants compiler level support for user experience, I think we should consider
instead implementing the the feature there perhaps and sidestepping many problems (while also introducing some serious other complications).

We can add more detailed errors, but I think it will make it jus more complicated.

TheMrAI committed:
feat(verifier): add verify_authenticate_param_type (#8295)

A suggestion to deal with `TypeParameter`. In particular, we can just
check that the type has no key ability, i.e., accept abilities just for
primitive types
https://github.com/iotaledger/iota/blob/dceb64b9ccf364b0438015b0cc3a1540bd9eeeb7/external-crates/move/crates/move-binary-format/src/file_format.rs#L850-L851

Moreover, I added new test cases. Is this now an exhaustive set of
tests?

TheMrAI committed:
refactor(iota-types): Add some much needed documentation for MovePackage

Some cases are still not covered, but this will massively help anybody
trying to understand what the code mean by storage/runtime/package/module
id and in what aspects do they morph into one another.

TheMrAI committed:
feat(iota-execution): Add RawModuleLoader

To properly be able to support create_auth_info we must be able
to load CompiledModules on the fly independent of the current
package or linkage context.
This is what this commit is supposed to provide, but for the time
being it could not be tested.

miker83z commented on crates/iota-types/src/base_types.rs:
```diff
@@ -132,6 +132,9 @@ pub trait ConciseableName<'a> {
 #[derive(Eq, PartialEq, Clone, Copy, PartialOrd, Ord, Hash, Serialize, Deserialize, JsonSchema)]
+/// `Storage ID` of any shared Object
+/// 
+/// The ID by which the object can be identified in global storage.
```
Storage ID as a specific context from which the object ID alone may be unrelated. I'd keep that out from the description of this type.
```suggestion
/// The ID by which the object can be identified in global storage.
```

TheMrAI replied:
Fixed in: 859e18f20cc0c938824bd54b1fc5524b77a37aae

miker83z commented on crates/iota-types/src/move_package.rs:
```diff
@@ -2,6 +2,37 @@
+//! The code contains terminology that may be confusing for the uninitiated,
+//! like `Module ID`, `Package ID`, `Storage ID` and `Runtime ID`. For avoidance
+//! of doubt these concepts are defined as so:
+//! - `Package ID` is the [ObjectID] representing the address by which the given
+//!   package may be found in storage.
```
Maybe this to make it different from the storage id?
```suggestion
//! - `Package ID` is the [ObjectID] representing the address of an immutable object which
//!   contains as data the package bytes.
```

TheMrAI replied:
Fixed in: 859e18f20cc0c938824bd54b1fc5524b77a37aae
Well, it is the same as `Package ID`, so instead I highlighted it more. 
Unless I misunderstood you somewhere.

miker83z commented on crates/iota-types/src/move_package.rs:
```diff
@@ -2,6 +2,37 @@
+//!   package may be found in storage.
+//! - `Storage ID` represents the [ObjectID] by which an object may be found in
+//!   storage. Be that a [MovePackage], [CompiledModule] or just any shared
+//!   object.
```
I guess also immutable and owned objects.
```suggestion
//!   storage. Be that a [MovePackage], [CompiledModule] or just any object.
```

TheMrAI replied:
Fixed in: 859e18f20cc0c938824bd54b1fc5524b77a37aae
Moved them into ObjectId for even better docs.

miker83z commented on crates/iota-types/src/move_package.rs:
```diff
@@ -2,6 +2,37 @@
+//! - `Runtime ID` will always mean the `Package ID`/`Storage ID` of the
+//!   initially published package. For a non upgradeable package this will
+//!   always be the `Storage ID`. For an upgradeable package, it will be the
+//!   `Storage ID` of the package's first deployed version.
```
```suggestion
//!   always be equal to the `Storage ID`. For an upgradeable package, it will be equal to the
//!   `Storage ID` of the package's first deployed version.
```

TheMrAI replied:
Fixed in: 859e18f20cc0c938824bd54b1fc5524b77a37aae

TheMrAI committed:
refactor(iota-types): Adjust documentation for MovePackages

miker83z commented on crates/iota-types/src/base_types.rs:
```diff
@@ -130,6 +130,11 @@ pub trait ConciseableName<'a> {
 
+/// The ID by which the object can be identified in global storage
+///
+/// Be that a [MovePackage](crate::move_package::MovePackage),
+/// [CompiledModule](move_binary_format::CompiledModule) or just
+/// any shared/immutable object.
```
This is mixing ownership and object data type. [Object `Data`](https://github.com/iotaledger/iota/blob/005587cead6cbfe071886e1acecd36f42b10d743/crates/iota-types/src/object.rs#L373) can be `MoveObject` or `MovePackage`. [Object `Owner`](https://github.com/iotaledger/iota/blob/005587cead6cbfe071886e1acecd36f42b10d743/crates/iota-types/src/object.rs#L450) can be `AddressOwner`, `ObjectOwner`, `Shared`, `Immutable`.

But I don't see why specifying those object properties for the ObjectID.

TheMrAI replied:
It is not intending to mix anything, but help place this actual type into a mental model.
When one says Object that can mean a lot of things. An instance of a Struct is an object, regardless of the Move language. Even just a piece of memory can be described as an object if you know it represents a type.

This sentence is highlighting the conceptual connection to the reader that the Object that they read about in external Move docs, describing that a package is an object is the same as this MovePackage. 
I think a CompiledModule is not an Object now that I think about it.

| or any shared/immutable object
Is not focusing on the ownership, just expanding the scope of the sentence.

Here one specific "package"  example and btw this also means in general "any shared/immutable object".

I also notice the sentence was a little hastily moved.
How about this:
```
/// The ID by which the `Move object` can be identified in global storage
///
/// This could be a [MovePackage](crate::move_package::MovePackage) or
/// any other shared/immutable `Move object`.
```
But I can remove `shared/immutable` part etc. Not sure how it would make more sense.

miker83z replied:
How about this?
```
/// The ID by which the `Move object` can be identified in global storage
///
/// This could be a [MovePackage](crate::move_package::MovePackage) or
/// any other [MoveObject](crate::object::MoveObject).
```

TheMrAI replied:
Fixed in: 5b0592dc47e2d9e289a1e4a956b3609a89f7b841

TheMrAI committed:
feat(iota-framework): Add create_auth_info_self_v1 implementation

This function relies on the currently active stack frame from the
Interpreter for identifying in which package it is in.
A few preconditions are used here. First we know that within a
given package all module references are resolved during runtime.
The stack frames contain the runtime id's of each module, which
together with the resolved package linkage uniquely identify
the module source. Allowing the function to reclaim the appropriate
CompiledModule.

To allow this with minimal modification DataStorage was propagated
into NativeContext.
This makes the code more confusing, because the Loader does not store
DataStorage at all, but always receives it.
If we decide there is no need for create_auth_info_self, then we
can remove this dependency.

TheMrAI committed:
refactor(iota-types): Fix ObjectID comment

TheMrAI committed:
refactor(iota-types): AuthenticatorInfoV1 now follows expected type check patterns

TheMrAI committed:
refactor(iota-execution): Fix fmt/clippy issues

TheMrAI committed:
refactor(iota-execution): Fix dprint and license checks

TheMrAI committed:
refactor(iota-execution): Fix rustdoc misunderstanding

TheMrAI committed:
fix(iota-protocol-config): Update it to include gas cost for create_auth_info_self

Dkwcs commented on iota-execution/latest/iota-move-natives/src/account.rs:
```diff
@@ -0,0 +1,172 @@
+        account_auth_verifier::verify_authenticate_func(&compiled_module, function_identifier)
+    {
+        return Err(
+            PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
```
I guess in future we have to standardize error codes for the main failure cases - for instance(E_MODULE_NOT_FOUND, E_FUNC_NOT_FOUND, E_SIG_MISMATCH, etc)
Just reasoning

TheMrAI replied:
TBH I forgot about the Error, just put in something to have it. Have to double check if there is something more appropriate.

Dkwcs commented on iota-execution/latest/iota-move-natives/src/account.rs:
```diff
@@ -0,0 +1,172 @@
+        account_create_auth_info_v1_impl_params.create_auth_info_v1_cost_base
+    );
+
+    let function_name_bytes = pop_arg!(args, VectorRef);
```
Just a refactoring suggestion we can extract a single or few functions that encapsulate resolution + validation

```
context
        .extensions_mut()
        .get::<NativesCostTable>()
        .account_create_auth_info_v1_impl_params;
    native_charge_gas_early_exit!(
        context,
        account_create_auth_info_v1_impl_params.create_auth_info_v1_cost_base
    );

    let function_name_bytes = pop_arg!(args, VectorRef);
    
    let function_name = String::from(unsafe {
        std::str::from_utf8_unchecked(function_name_bytes.as_bytes_ref().as_slice())
    });
    let function_identifier = Identifier::new(function_name.clone()).unwrap();

    let module_name_bytes = pop_arg!(args, VectorRef);
    let module_name = String::from(unsafe {
        std::str::from_utf8_unchecked(module_name_bytes.as_bytes_ref().as_slice())
    });
    let module_identifier = Identifier::new(module_name.clone()).unwrap();

    let package = pop_arg!(args, AccountAddress);
    let package_id = ObjectID::from(package);
```

etc

in order to have something like 

```
struct AuthenticateData<‘a> {
    module: CompiledModule,
    func_handle: FunctionHandleIndex,
    is_public_entry: bool,
    sig_ok: bool,
}

fn resolve_auth_data(
    ctx: &mut NativeContext,
    package_id: Option<ObjectID>,    // None => derive from active frame (self)
    module_name: &str,
    function_name: &str,
) -> VMResult<AuthenticateData<'_>> {
    let (pkg_id, module_ident) = resolve_package_and_module(ctx, package_id, module_name)?;
    let module = load_compiled_module(ctx, pkg_id, &module_ident)?; // via RawModuleLoader or active frame
    let (fh, vis) = lookup_function(&module, function_name)?;
    enforce_public_entry(vis)?;
    enforce_authenticate_signature(&module, fh)?;
    Ok(ResolvedAuthenticate { module, func_handle: fh, is_public_entry: true, sig_ok: true })
}

```

TheMrAI replied:
Good points! I was hesitant at this stage in refactoring as I am still unsure if it would actually work.

TheMrAI commented:
This PR will be split, do not review.
